subprojects/security/src/main/java/org/gradle/security/internal/PublicKeyDownloadService.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.security.internal;
+
+import org.bouncycastle.openpgp.PGPObjectFactory;
+import org.bouncycastle.openpgp.PGPPublicKey;
+import org.bouncycastle.openpgp.PGPPublicKeyRing;
+import org.bouncycastle.openpgp.PGPUtil;
+import org.bouncycastle.openpgp.operator.bc.BcKeyFingerprintCalculator;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.internal.io.ExponentialBackoff;
+import org.gradle.internal.resource.transfer.ExternalResourceAccessor;
+import org.gradle.internal.resource.transfer.ExternalResourceReadResponse;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import static org.gradle.security.internal.SecuritySupport.toHexString;
+
+public class PublicKeyDownloadService implements PublicKeyService {
+    private final static Logger LOGGER = Logging.getLogger(PublicKeyDownloadService.class);
+
+    private final List<URI> keyServers;
+    private final ExternalResourceAccessor client;
+
+    public PublicKeyDownloadService(List<URI> keyServers, ExternalResourceAccessor client) {
+        this.keyServers = keyServers;
+        this.client = client;
+    }
+
+    @Override
+    public Optional<PGPPublicKey> findPublicKey(long id) {
+        List<URI> servers = new ArrayList<>(keyServers);
+        Collections.shuffle(servers);
```

Comment:
Why the `shuffle`?

Reply:
This is to avoid spamming the same server again and again.